src/utils/read-rows-resumption.ts

```diff
@@ -0,0 +1,267 @@
+        range.end as BoundData,
+        'Key'
+      )
+    );
+
+    if (this.options.filter) {
+      reqOpts.filter = Filter.parse(this.options.filter);
+    }
+
+    if (this.hasLimit) {
+      reqOpts.rowsLimit = this.rowsLimit - this.rowsRead;
+    }
+    return reqOpts;
+  }
+
+  /**
+   * Decides if the client is going to retry a request.
+   *
+   * canResume contains the logic that will decide if the client will retry with
+   * another request when it receives an error. This logic is passed along to
+   * google-gax and used by google-gax to decide if the client should retry
+   * a request when google-gax receives an error. If canResume returns true then
+   * the client will retry with another request computed by getResumeRequest. If
+   * canResume request returns false then the error will bubble up from gax to
+   * the handwritten layer.
+   *
+   * @param {GoogleError} [error] The error that Google Gax receives.
+   * @return {boolean} True if the client will retry
+   */
+  canResume(error: GoogleError): boolean {
+    // First update the row keys and the row ranges based on the last row key.
+    const lastRowKey = this.chunkTransformer
+      ? this.chunkTransformer.lastRowKey
+      : '';
+    if (lastRowKey) {
+      spliceRanges(this.ranges, lastRowKey);
+      this.rowKeys = getRowKeys(this.rowKeys, lastRowKey);
+    }
+
+    // If all the row keys and ranges are read, end the stream
+    // and do not retry.
+    if (this.rowKeys.length === 0 && this.ranges.length === 0) {
+      return false;
+    }
+    // If there was a row limit in the original request and
+    // we've already read all the rows, end the stream and
+    // do not retry.
+    if (this.hasLimit && this.rowsLimit === this.rowsRead) {
+      return false;
+    }
+    const isRstStreamError = (error: GoogleError | ServiceError): boolean => {
+      // Retry on "received rst stream" errors
+      if (error.code === 13 && error.message) {
+        const error_message = (error.message || '').toLowerCase();
+        return (
+          error.code === 13 &&
+          (error_message.includes('rst_stream') ||
+            error_message.includes('rst stream'))
+        );
+      }
+      return false;
+    };
+    if (
+      error.code &&
+      (RETRYABLE_STATUS_CODES.has(error.code) || isRstStreamError(error))
+    ) {
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * Creates a RetryOptions object that can be used by google-gax.
+   *
+   * This class contains the business logic to specify retry behaviour of
+   * readrows requests and this function packages that logic into a RetryOptions
+   * object that google-gax expects.
+   *
+   * @param {CallOptions} [gaxOpts] The call options that will be used to
+   * specify retry behaviour.
+   * @return {RetryOptions} A RetryOptions object that google-gax expects that
+   * can determine retry behaviour.
+   *
+   */
+  toRetryOptions(gaxOpts: CallOptions): RetryOptions {
```

Comment:
No actual notes here, it's just really cool to see this application of the gax stuff finally!!!